src/main/scala/org/constellation/snapshotstreaming/FileBasedLastIncrementalGlobalSnapshotStorage.scala

```diff
@@ -0,0 +1,121 @@
+import org.tessellation.security.Hashed
+
+import fs2.io.file._
+import fs2.{Stream, text}
+import io.circe.Codec
+import io.circe.generic.semiauto.deriveCodec
+import io.circe.parser.decode
+import io.circe.syntax._
+
+object FileBasedLastIncrementalGlobalSnapshotStorage {
+
+  def make[F[_]: Async: Files: KryoSerializer](
+    path: Path
+  ): F[LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo]] =
+    Semaphore[F](1)
+      .map(make(path, _))
+
+  private def make[F[_]: Async: Files: KryoSerializer](
+    path: Path,
+    semaphore: Semaphore[F]
+  ): LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo] =
+    new LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo] {
+
+      implicit val codec: Codec[Hashed[GlobalIncrementalSnapshot]] = deriveCodec[Hashed[GlobalIncrementalSnapshot]]
+      implicit val snapshotWithInfoCodec: Codec[SnapshotWithState] = deriveCodec[SnapshotWithState]
+
+      def set(snapshot: Hashed[GlobalIncrementalSnapshot], state: GlobalSnapshotInfo): F[Unit] =
+        semaphore.permit.use { _ =>
+          validateStateProof(snapshot, state) >>
+            get.flatMap {
+              case Some(last) if isNextSnapshot(last, snapshot.signed.value) =>
+                Stream
+                  .emit(SnapshotWithState(snapshot, state).asJson.spaces2)
+                  .through(text.utf8.encode)
+                  .through(Files[F].writeAll(path))
+                  .compile
+                  .drain
+              case Some(last) =>
+                MonadThrow[F].raiseError[Unit](
+                  new Throwable(
+                    s"Snapshot is not the next one! last: ${SnapshotReference
+                        .fromHashedSnapshot(last)}, lastHash: ${last.hash}, next: ${SnapshotReference
+                        .fromHashedSnapshot(snapshot)}, prevHash: ${snapshot.signed.value.lastSnapshotHash}"
+                  )
+                )
+              case None =>
+                MonadThrow[F].raiseError[Unit](
+                  new Throwable("Previous snapshot not found when setting next global snapshot!")
+                )
+            }
+        }
+
+      def setInitial(snapshot: Hashed[GlobalIncrementalSnapshot], state: GlobalSnapshotInfo): F[Unit] =
+        semaphore.permit.use { _ =>
+          validateStateProof(snapshot, state) >>
+            Stream
+              .emit(SnapshotWithState(snapshot, state).asJson.spaces2)
+              .through(text.utf8.encode)
+              .through(Files[F].writeAll(path, Flags(Flag.Write, Flag.CreateNew)))
+              .compile
+              .drain
+        }
+
+      private def validateStateProof(snapshot: Hashed[GlobalIncrementalSnapshot], state: GlobalSnapshotInfo) =
+        state
+          .stateProof[F]
+          .map(_ === snapshot.stateProof)
+          .flatMap((new Throwable("State proof doesn't match!").raiseError[F, Unit].unlessA))
```

Comment:
I understand that this is a critical error, right? A snapshot with invalid state proof should never make it here unless there would be an implementation error? I wonder about whether we should throw or not.